### PR TITLE
haptic on/off setting

### PR DESCRIPTION
**Pull request by TychoVrahe**
Adds setting to turn haptics on and off.

### CODE REVIEW

File: core/embed/extmod/modtrezorutils/modtrezorutils.c [hunk 1]
```diff
@@ -439,7 +441,12 @@ STATIC const mp_rom_map_elem_t mp_module_trezorutils_globals_table[] = {
 #ifdef USE_BACKLIGHT
     {MP_ROM_QSTR(MP_QSTR_USE_BACKLIGHT), mp_const_true},
 #else
-    {MP_ROM_QSTR(MP_QSTR_USE_BACKLIGHT), mp_const_false},
+        {MP_ROM_QSTR(MP_QSTR_USE_BACKLIGHT), mp_const_false},
```
Comment (prusnak) [inline core/embed/extmod/modtrezorutils/modtrezorutils.c:444]: unnecessary reformat?

File: core/embed/extmod/modtrezorutils/modtrezorutils.c [hunk 2]
```diff
@@ -439,7 +441,12 @@ STATIC const mp_rom_map_elem_t mp_module_trezorutils_globals_table[] = {
...
+#endif
+#ifdef USE_HAPTIC
+    {MP_ROM_QSTR(MP_QSTR_USE_HAPTIC), mp_const_true},
+#else
+        {MP_ROM_QSTR(MP_QSTR_USE_HAPTIC), mp_const_false},
```
Comment (prusnak) [inline core/embed/extmod/modtrezorutils/modtrezorutils.c:449]: unnecessary whitespace?

**Review by prusnak — CHANGES_REQUESTED** [review]

File: core/embed/extmod/modtrezorutils/modtrezorutils.c (hunk 1 above)
Comment (TychoVrahe) [reply core/embed/extmod/modtrezorutils/modtrezorutils.c]: fixed https://github.com/trezor/trezor-firmware/pull/3802/commits/aa1f7b5a394d1286e44518cf8d660455bca9733a

File: core/embed/extmod/modtrezorutils/modtrezorutils.c (hunk 2 above)
Comment (TychoVrahe) [reply core/embed/extmod/modtrezorutils/modtrezorutils.c]: fixed https://github.com/trezor/trezor-firmware/pull/3802/commits/aa1f7b5a394d1286e44518cf8d660455bca9733a

File: core/embed/extmod/modtrezorio/modtrezorio.c [hunk 3]
```diff
@@ -55,9 +55,12 @@ bool usb_connected_previously = true;
 #include "modtrezorio-fatfs.h"
 #include "modtrezorio-sdcard.h"
 #endif
+#ifdef USE_HAPTIC
+#include "modtrezorio-haptic.h"
+#endif
 
 /// package: trezorio.__init__
-/// from . import fatfs, sdcard
+/// from . import fatfs, sdcard, haptic
```
[the remark this answers is not in the document]
Comment (TychoVrahe) [reply core/embed/extmod/modtrezorio/modtrezorio.c]: fixed https://github.com/trezor/trezor-firmware/commit/4f700c68cdc9f0452555bf40a49023c4b64f7614

File: core/mocks/generated/trezorio/__init__.pyi [hunk 4]
```diff
@@ -190,7 +190,7 @@ class WebUSB:
         """
         Sends message using USB WebUSB (device) or UDP (emulator).
         """
-from . import fatfs, sdcard
+from . import fatfs, sdcard, haptic
```
[the remark this answers is not in the document]
Comment (TychoVrahe) [reply core/mocks/generated/trezorio/__init__.pyi]: fixed https://github.com/trezor/trezor-firmware/commit/4f700c68cdc9f0452555bf40a49023c4b64f7614

File: core/src/boot.py [hunk 5]
```diff
@@ -10,7 +10,7 @@
 
 import storage
 import storage.device
-from trezor import config, log, loop, ui, utils, wire, translations
+from trezor import config, log, loop, ui, io, utils, wire, translations
```
[the remark this answers is not in the document]
Comment (TychoVrahe) [reply core/src/boot.py]: fixed https://github.com/trezor/trezor-firmware/commit/4f700c68cdc9f0452555bf40a49023c4b64f7614

isort is disabled for this file and from the others where was is an issue one is actually C file from which the third is generated, so no detection there either.

File: common/protob/messages-management.proto [hunk 6]
```diff
@@ -182,6 +184,7 @@ message ApplySettings {
     optional bool experimental_features = 10;  // enable experimental message types
     optional bool hide_passphrase_from_host = 11;  // do not show passphrase coming from host
     optional uint32 brightness = 12;  // display brightness, 0 = select on device
+    optional bool haptic_feedback = 13;  // enable haptic feedback
```
[the remark this answers is not in the document]
Comment (TychoVrahe) [reply common/protob/messages-management.proto:187]: could be, but comments for other boolean values here do not use that, so i think its better to keep the same style

File: core/src/storage/device.py [hunk 7]
```diff
@@ -36,6 +36,7 @@
 _EXPERIMENTAL_FEATURES     = const(0x15)  # bool (0x01 or empty)
 _HIDE_PASSPHRASE_FROM_HOST = const(0x16)  # bool (0x01 or empty)
 _BRIGHTNESS                = const(0x17)  # int
+_DISABLE_HAPTIC_FEEDBACK   = const(0x18)  # bool (0x01 or empty)
```
[the remark this answers is not in the document]
Comment (TychoVrahe) [reply core/src/storage/device.py:39]: The default is to have haptic feedback enabled - so when there is no value written is storage it needs to be interpreted as enabled too. 

Our boolean getter in storage/common returns `false` when no value is written, so false needs to be interpreted as enabled.

File: core/embed/extmod/modtrezorio/modtrezorio-haptic.h [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "haptic.h"
+
+/// package: trezorio.haptic
+
+/// def haptic_set(enable: bool) -> None:
```
[the remark this answers is not in the document]
Comment (TychoVrahe) [reply core/embed/extmod/modtrezorio/modtrezorio-haptic.h]: About the naming here, and applies to several comments below too.

There are two different things in question
`haptic driver` - the low level driver controlling the haptic driver IC and haptic motor, 
`haptic feedback` - a specific function realized by the haptic driver. But other functionality, like vibration alert, can be realized by haptic driver

This is what i tried to respect when i chose the names. 

The file here specifically is python interface for controlling the haptic driver and this set function currently disables all functionality realized by the haptic driver, not just haptic feedback, so i think the word feedback does not belong here. 

I modified the name function to match the underlying haptic driver name, so the new name is `haptic_set_enabled`

File: core/embed/trezorhal/haptic.h [hunk 9]
```diff
@@ -23,4 +23,6 @@ bool haptic_test(uint16_t duration_ms);
 // Play haptic effect
 void haptic_play(haptic_effect_t effect);
 
+void haptic_set_enabled(bool enable);
```
[the remark this answers is not in the document]
Comment (TychoVrahe) [reply core/embed/trezorhal/haptic.h:26]: See above for explanation why not include `feedback` here. 

I think `set_enabled` is better here - when we implement the set_intensity function i think its better to have clearer distinction what is actually being set. 

I renamed the `haptic_set` in https://github.com/trezor/trezor-firmware/commit/4f700c68cdc9f0452555bf40a49023c4b64f7614 so that we just have consisten, more descriptive names everywhere

File: core/mocks/generated/trezorutils.pyi [hunk 10]
```diff
@@ -119,6 +119,8 @@ USE_SD_CARD: bool
 """Whether the hardware supports SD card."""
 USE_BACKLIGHT: bool
 """Whether the hardware supports backlight brightness control."""
+USE_HAPTIC: bool
```
[the remark this answers is not in the document]
Comment (TychoVrahe) [reply core/mocks/generated/trezorutils.pyi:122]: See also above, this is meant to represent the presence of haptic driver.

File: core/embed/trezorhal/stm32u5/haptic/drv2625/drv2625.c [hunk 11]
```diff
@@ -191,3 +197,5 @@ void haptic_play(haptic_effect_t effect) {
 bool haptic_test(uint16_t duration_ms) {
   return haptic_play_RTP(PRODTEST_EFFECT_AMPLITUDE, duration_ms);
 }
+
+void haptic_set_enabled(bool enable) { haptic_enabled = enable; }
```
[the remark this answers is not in the document]
Comment (TychoVrahe) [reply core/embed/trezorhal/stm32u5/haptic/drv2625/drv2625.c:201]: See above

File: core/src/apps/management/apply_settings.py [hunk 12]
```diff
@@ -278,3 +287,22 @@ def callback(val: int) -> None:
             count=style.get_backlight_normal(),
             br_name="set_brightness",
         )
+
+
+if utils.USE_HAPTIC:
+
+    async def _require_confirm_haptic_feedback(enable: bool) -> None:
+        if enable:
```
[the remark this answers is not in the document]
Comment (TychoVrahe) [reply core/src/apps/management/apply_settings.py]: fixed https://github.com/trezor/trezor-firmware/commit/4f700c68cdc9f0452555bf40a49023c4b64f7614

using just one `br_type`

File: core/src/trezor/enums/Capability.py [hunk 13]
```diff
@@ -11,6 +11,7 @@
 PassphraseEntry = 17
 Translations = 19
 Brightness = 20
+Haptic = 21
```
[the remark this answers is not in the document]
Comment (TychoVrahe) [reply core/src/trezor/enums/Capability.py:14]: Also here i wanna indicate the presence of haptic driver, not just the more specific function of haptic feedback

File: core/src/storage/device.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (ibz) [reply core/src/storage/device.py:39]: Makes sense. Then I would have argued to rename everything else top match this, because indeed, we are adding an option to disable it, not to enable it. But this can also work...

**Review by mmilata — APPROVED** [review]

File: core/embed/rust/src/ui/model_mercury/flow/confirm_action.rs [hunk 14]
```diff
@@ -118,6 +118,7 @@ fn new_confirm_action_obj(_args: &[Obj], kwargs: &Map) -> Result<Obj, error::Err
     let title: TString = kwargs.get(Qstr::MP_QSTR_title)?.try_into()?;
     let action: Option<TString> = kwargs.get(Qstr::MP_QSTR_action)?.try_into_option()?;
     let description: Option<TString> = kwargs.get(Qstr::MP_QSTR_description)?.try_into_option()?;
+    let subtitle: Option<TString> = kwargs.get(Qstr::MP_QSTR_subtitle)?.try_into_option()?;
```
Comment (matejcik) [inline core/embed/rust/src/ui/model_mercury/flow/confirm_action.rs:121]: this is causing the FIDO failure
because you require the `subtitle` argument (though it is allowed to be `None`, it must be present) and `confirm_fido_reset()` is not setting it.

the call must be uhh
```suggestion
    let subtitle: Option<TString> = kwargs.get(Qstr::MP_QSTR_subtitle)?.unwrap_or(Obj::const_none()).try_into_option()?;
```
